### PR TITLE
Test the "game" module

### PR DESCRIPTION
**Pull request by girardb**

### CODE REVIEW

File: src/game/game.py [hunk 1]
```diff
@@ -70,27 +88,36 @@ def get_game_state(self):
         return json.dumps(game_state)
 
     def bleed_players(self):
-        for player in self._players:
-            player.bleed()
+        for username in self._players:
```
Comment (girarda) [inline src/game/game.py]: `for players in self._players.values()` ?

File: src/game/game.py [hunk 2]
```diff
@@ -70,27 +88,36 @@ def get_game_state(self):
...
+        self._players[username].click()
+
     def tied_game(self):
         print('everyone is a loser :(')
         self.ongoing = False
 
     def income_tick(self):
-        for player in self._players:
-            player.income_tick()
+        for username in self._players:
```
Comment (girarda) [inline src/game/game.py]: `for players in self._players.values()` ?

File: src/game/game.py [hunk 3]
```diff
@@ -70,27 +88,36 @@ def get_game_state(self):
...
     def game_won(self):
         pass
 
+    def click(self, username):
+        if username not in self._players:
+            raise NonExistentUserException("This user is not in the current game.")
+        self._players[username].click()
+
     def tied_game(self):
         print('everyone is a loser :(')
```
Comment (girarda) [inline src/game/game.py]: lol

File: src/game/game.py [hunk 4]
```diff
@@ -46,22 +48,38 @@ def single_tick(self):
...
+        if not players_alive:
+            self.tied_game()
+        else:
+            pass
+
     def _tick(self):
         self.single_tick()
         players_alive = self.players_alive()
-        if len(players_alive) >= 1: # Actual line -> len(players_alive > 1)
+        if len(players_alive) > 1:
```
Comment (girarda) [inline src/game/game.py]: Shouldn't this check if the game is ongoing?

File: src/game/player.py [hunk 5]
```diff
@@ -1,19 +1,27 @@
 class Player:
-    def __init__(self):
+    def __init__(self, name='unnamed'):
```
Comment (girarda) [inline src/game/player.py]: What's the purpose of the default parameter? In what context do we want an unnamed Player?

File: src/game/driver.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+class ProdDriver(Driver):
+    """
+    Runs 'real' games
+    """
+    def re_tick(self):
+        tick = threading.Timer(1.0, self._tick)
+        tick.start()
+
+
+class TestDriver(Driver):
```
Comment (girarda) [inline src/game/driver.py]: Can probably move this to the test class. this is only for testing so it doesn't have to be in the actual source.

File: src/game/game.py [hunk 7]
```diff
@@ -6,66 +6,48 @@
...
 
     def add_server(self, server):
         self.server = server
 
-    def pregame_setup(self):
-        if not self._players:
-            raise EmptyGameException("The game lobby is empty.")
+    def _pregame_setup(self):
+        if len(self._players) < 2:
+            raise EmptyGameException("The game lobby is too empty.")
```
Comment (girarda) [inline src/game/game.py]: `Not enough players`? Too empty sounds weird, but I won't fight against it

File: src/game/driver.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        self._tick()
+
+    def re_tick(self):
+        raise NotImplementedError()
+
+    def _tick(self):
+        if not self.game.is_over():
+            self.game.single_tick()
+            self.re_tick()
+        elif self.game.is_over() and not self.game.players_alive():
```
Comment (girarda) [inline src/game/driver.py]: I'd move that logic to the game. How about `single_tick` returns a `GameState`? That way the driver knows when to stop, but it doesn't have to handle the logic of calling `tied_game` or `game_won`

File: src/game/player.py [hunk 9]
```diff
@@ -1,12 +1,13 @@
...
-    def __init__(self, name='unnamed'):
+    def __init__(self, name):
         self.hp = 100
         self.income = 1
         self.cookies = 0
         self.click_value = 1
         self.bleed_amount = 20
         self.total_damage_dealt = 0
         self.name = name
+        self.token = ""
```
Comment (girarda) [inline src/game/player.py]: What is token used for?

File: src/game/tests/test_game.py [hunk 10]
```diff
@@ -9,13 +9,13 @@ def setUp(self) -> None:
         self.game = Game()
 
     def test_no_players(self):
-        self.assertRaises(EmptyGameException, self.game.pregame_setup)
+        self.assertRaises(EmptyGameException, self.game._pregame_setup)
```
Comment (girarda) [inline src/game/tests/test_game.py]: You probably want to add a test with only one player since the boundary value changed

File: src/game/tests/test_driver.py [hunk 11]
```diff
@@ -9,6 +10,21 @@ def setUp(self) -> None:
         game = Game()
         self.driver = TestDriver(game)
 
+    def test_game_over(self):
+        create_players(self.driver.game, 3)
+        self.driver.start_game()
+
+        self.assertEqual(self.driver.game.players_alive(), [])
+
```
Comment (girardb) [inline src/game/tests/test_driver.py]: Forgot to assertFalse(self.driver.game.ongoing)
